sdk/plugin/backend.go

```diff
@@ -24,18 +25,33 @@ type GRPCBackendPlugin struct {
 	MetadataMode bool
 	Logger       log.Logger
 
+	MultiplexingSupport bool
+
 	// Embeding this will disable the netRPC protocol
 	plugin.NetRPCUnsupportedPlugin
 }
 
 func (b GRPCBackendPlugin) GRPCServer(broker *plugin.GRPCBroker, s *grpc.Server) error {
-	pb.RegisterBackendServer(s, &backendGRPCPluginServer{
-		broker:  broker,
-		factory: b.Factory,
-		// We pass the logger down into the backend so go-plugin will forward
-		// logs for us.
+	server := backendGRPCPluginServer{
+		broker:    broker,
+		factory:   b.Factory,
+		instances: make(map[string]backendInstance),
+		// We pass the logger down into the backend so go-plugin will
+		// forward logs for us.
 		logger: b.Logger,
-	})
+	}
+
+	if b.MultiplexingSupport {
+		// Multiplexing is enabled for this plugin, register the server so we
+		// can tell the client in Vault.
+		pluginutil.RegisterPluginMultiplexingServer(s, pluginutil.PluginMultiplexingServerImpl{
+			Supported: true,
+		})
+	} else {
+		server.instances["single"] = backendInstance{}
```

Comment:
Should this ID, `"single"`, be a `const`? This is an interesting approach compared to what we have on the Database side, where we keep the `singleImpl` as a top-level field.

Reply:
Do we have to set this entry on the map here, or would `Setup` handle this? It doesn't look like we populate `backendInstance` so I'm curious why we need to have this instance entry set.

Reply:
Yes, this approach is different than DB plugins. I am happy to use a similar pattern to DBs if you think that makes more sense.

> Do we have to set this entry on the map here

The `backendGRPCPluginServer.Setup()` call needs to know if we are multiplexed or not because it has to get the multiplexed ID from the context. So we tell `Setup` that this backendGRPCPluginServer is multiplexed by setting the `backendGRPCPluginServer.instances["single"] = backendInstance{}`. [Then in `Setup` we always look for `single` first](https://github.com/hashicorp/vault/pull/14946/files#diff-a5bab53bc70ad6f4380960cdf037117390a05cc089a8d77c21e697eb253c8e2eR94). If it isn't there we [call getMultiplexIDFromContext](https://github.com/hashicorp/vault/pull/14946/files#diff-a5bab53bc70ad6f4380960cdf037117390a05cc089a8d77c21e697eb253c8e2eR95).